File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        mp = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz",
        }

        res = []

        # process
        for digit in digits:
            if len(res) == 0:
                res = mp[digit]
                continue
            
            arr = mp[digit]
            newres = []
            for x in res:
                for y in arr:
                    newres.append(x+y)
            
            res = newres

        return res

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        map<char, string> mp = {
            {'2',"abc"},
            {'3',"def"},
            {'4',"ghi"},
            {'5',"jkl"},
            {'6',"mno"},
            {'7',"pqrs"},
            {'8',"tuv"},
            {'9',"wxyz"},
        };

        vector<string> res;

        for (auto& digit: digits) {
            if (res.size() == 0) {
                for (auto& ch: mp[digit]) {
                    res.push_back(string(1,ch));
                }
                continue;
            }

            auto arr = mp[digit];
            vector<string> newres;
            for (auto& x:res) {
                for (auto& y:arr) {
                    newres.push_back(x+y);
                }
            }

            res = newres;
        }

        return res;
    }
};